### ISSUE TITLE
Native Chromecast support

### ISSUE THREAD

**Issue by AvdWerfhorst** [deluan/navidrome#250]
First of all, I want to say, since last weekend my SD card in my Raspberry Pi crashed, I switched from Airsonic to Navidrome. And I really like it so far. I noticed that I had some discrepancies in my tagging, so that made me edit a bit of metadata in my files. Takes some time, but in the end improves my music library.

There is a functionality that I would love to see. Native support for Chromecast. In my current situation I use Dsub on my phone, to connect to one of my Chromecast audio devices (or to a group if I go multiroom). What works great and is also visible as an extra connection in the web interface. But it is pretty clear that my phone is the device feeding the Chromecast with information what flac file to play. As when I walk out of my Wifi reach and the song ends, the next song does not start.

To me it would feel logic to get these controls on the server side. So we have the device that acts as source for the flac files, feed the information to the Chromecast.

Saying that, of course there are at the moment bigger priorities. And the way it works now, it works great.

**Comment by deluan** [comment]
I also want to implement this, personally. But this needs to wait for the Jukebox implementation to be in place.

**Comment by brianpierson2020** [comment]
This would be great. With my home automation in mind (using home-assistant.io) it would be great if we expose part of this functionality if possible to the API as well. Like an API call to Play Song/Artist/Album/Playlist on Device X. That would be totally amazing.

**Comment by deluan** [comment]
Yes, when talking about Jukebox, I basically mean "Subsonic API's Jukebox": http://www.subsonic.org/pages/api.jsp#jukeboxControl

**Comment by Pollinopops** [comment]
Hey, I'm following Navidrome's developpment very closely. It looks really promissing ! For the moment, I am sticking with airsonic because I mainly listen to my music with the jukebox mode on my pi with a hifiberry DAC. Any ides when the feature will be implemented in Navidrome ?

Thanks for the great work !

**Comment by deluan** [comment]
@Pollinopops Do you mean the Jukebox mode #364 or Chromecast support? Or both? Jukebox will come first, hopefully in the next couple of months. Chromecast support may take longer, not sure how it works.

**Comment by Pollinopops** [comment]
Oh sorry I just realized I was in the wrong place... Yes I meant jukebox mode like in airsonic. Great news thanks ! Can't wait to try it out!

**Comment by Tropaion** [comment]
Are there any news about this feature? I would be really interested in it.

**Comment by BloodyIron** [comment]
This one feature missing alone means I won't consider using this suite. Chromecast devices are everywhere...

**Comment by berserkwarwolf** [comment]
> This one feature missing alone means I won't consider using this suite. Chromecast devices are everywhere...

I'm on the same boat. I need an audio server with CC support.

**Comment by TaakoMagnusen** [comment]
any movement on this?

**Comment by tjukkie** [comment]
Would love the support of chromecast.

**Comment by TaakoMagnusen** [comment]
any movement on this?

**Comment by codeinabox** [comment]
Is this even technically possible? Even the Spotify web app doesn't offer Chromecast support, for that you need to download either the Android or macOS desktop application

**Comment by kevincox** [comment]
It is possible with Chome. It provides a proprietary API to the page which can be used to cast content. There is a standardized version of the API in the works (I can't find it right now it was something like Remote Display API, Alternate Display API, Presentation API or something) but it is moving very slowly and not available in any browser.

**Comment by codeinabox** [comment]
Ah I see, do you mean the [Google Cast Web Sender API](https://developers.google.com/cast/docs/web_sender/integrate)?

**Comment by kevincox** [comment]
Actually I think I misunderstood what you are asking. I think what you are referencing is the proprietary API available in Chrome. That is not the standardized API.

**Comment by kevincox** [comment]
Yes! I believe that is it. Thanks for finding it. IDK how I couldn't locate the spec with a search.

**Comment by kevincox** [comment]
There is actually an MDN page for it as well and apparently it is already supported by Chromium-based browsers so likely it is the best target to build towards.

https://developer.mozilla.org/en-US/docs/Web/API/Presentation_API

Although neither Mozilla or Apple have published an official position so IDK if it will be adopted elsewhere.

**Comment by shagr4th** [comment]
fwiw, I started to work on this. While adding a chromecast session to the UI is quite simple (I already did add support on another web app), to make a nice UX of it is harder, as there's a need for low ui interaction with the react-music-player library... for instance to simulate (play/seek/stop/etc.) a "virtual track", mirroring the one actually playing on the chromecast

Two solutions:
- work on the [fork of react-music-player](https://github.com/navidrome/react-music-player) made by deluan and add chromecast support to it. But it's clearly stated "This fork won't be maintained"
- rewrite the whole music player

**Comment by Miriup** [comment]
> I started to work on this.

Since I got an old Chromecast adapter gifted for my birthday I'm also interested in this. @shagr4th  Where can I take a look at the work you already did?

Most likely for a rewrite a lot of work would be required. Personally I'd favor your suggestion to stick it into the fork of react-music-player. There's also a number of hangs I experience on my player that I want to investigate and fix as well.

**Comment by shagr4th** [comment]
I got nothing really good to share, I just added the chromecast button and session code, before getting stuck with the choice I wrote earlier. But, you can look at another subsonic web client called [subplayer](https://github.com/shagr4th/subplayer) I forked some years ago. There's a branch with added chromecast support. It's a little hacky and not bug free, but useable for managing music on chromecast from your laptop

**Comment by codeinabox** [comment]
Chromecast support via Music Assistant is an option as it already supports playing over Chromecast, and there is a [branch adding a Subsonic provider](https://github.com/orgs/music-assistant/discussions/414#discussioncomment-8150677) that I've been helping test against Navidrome.

**Comment by codeinabox** [comment]
The Subsonic provider has been merged into the Music Assistant server if anyone wants to help test it further https://github.com/music-assistant/server/releases/tag/2.0.0b84

**Comment by kevincox** [comment]
relevant

**Comment by michaelkrieger** [comment]
Relevant

**Comment by tcurdt** [comment]
I'll be honest: I once wanted this. But given that both AirPlay and Chromecast are locked down, I find a different approach now more appealing. Be it snapcast or a navidrome in proxy mode, both approaches seem easier and more open.

https://github.com/navidrome/navidrome/discussions/3539

**Comment by michaelkrieger** [comment]
> AirPlay and Chromecast are locked down

What does that mean exactly?

Right now I use a client on my phone, airplay to my amplifier in the living room.  The phone is the dependency. Needs to stay within range.  Presumably a loss in quality. No 24bits/96000Hz audio.

Direct chrome cast would allow me to play from navidrome to the amplifier without any quality loss and middle-man. That is the goal. 

[https://github.com/muammar/mkchromecast/wiki/FAQ](mkchromecast) will actually link into ALSA/PulseAudio and do it, but an interface is needed.  I guess this is the current workaround for a more permanent setup.

**Comment by tcurdt** [comment]
> > AirPlay and Chromecast are locked down
> 
> What does that mean exactly?

Airplay 1 allows for sending
Airplay 2 needs a key for sending
Airplay needs a key for receiving (some implementations extracted some keys)

I might be wrong on that but IIUC it's a similar situation for CC.

https://github.com/tristanpenman/chromecast-tools

If you just care about playing to your amplifier this might not matter.
But it still might matter to some people for other reasons.

And while it would still be nice to have CC support, I'd be now even more interested in other approaches.
That's all I wanted to add and point to the discussion of other options.

Here is some code that might be another starting point for a native CC support. 

https://github.com/vishen/go-chromecast

**Comment by sschnabe** [comment]
relevant
